loadbalancer/algorithm_test.go

```diff
@@ -335,10 +336,59 @@ func TestConsistentHashKey(t *testing.T) {
 	}
 
 	for i, ep := range endpoints {
-		key := ep // key equal to endpoint has the same hash and therefore selects it
+		key := fmt.Sprintf("%s-%d", ep, 1) // key equal to endpoint has the same hash and therefore selects it
 		selected := ch.Apply(&routing.LBContext{Request: r, Route: rt, Params: map[string]interface{}{ConsistentHashKey: key}})
 		if selected != rt.LBEndpoints[i] {
 			t.Errorf("expected: %v, got %v", rt.LBEndpoints[i], selected)
 		}
 	}
 }
+
+func TestConsistentHashKeyDistribution(t *testing.T) {
+	endpoints := []string{"http://127.0.0.1:9001", "http://127.0.0.1:9002", "http://127.0.0.1:9003", "http://127.0.0.1:9004", "http://127.0.0.1:9005", "http://127.0.0.1:9006", "http://127.0.0.1:9007", "http://127.0.0.1:9008", "http://127.0.0.1:9009", "http://127.0.0.1:9010"}
+	counters := map[string]int{}
+
+	route := NewAlgorithmProvider().Do([]*routing.Route{{
+		Route: eskip.Route{
+			BackendType: eskip.LBBackend,
+			LBAlgorithm: ConsistentHash.String(),
+			LBEndpoints: endpoints,
+		},
+	}})[0]
+	ch := route.LBAlgorithm.(consistentHash)
+	for i := range [100_000]int{} {
+		ctx := makeCtx(route, i)
+		selected := ch.Apply(ctx)
+		counters[selected.Host] += 1
+	}
+	counts := make([]int, 0, len(counters))
+	sum := 0
+	for _, v := range counters {
+		sum += v
+		counts = append(counts, v)
+	}
+	if sum != 100_000 {
+		t.Errorf("Sent 100,000 requests so should have selected 100,000 endpoints, got %d", sum)
+	}
+	stdDev := stdDeviation(counts, sum)
+	if stdDev >= 56 {
```

Comment:
Previous algorithm had standard deviation of 280% for these 10 endpoints

Reply:
I am not sure if this would be a source of test flakiness. An alternative approach could be to run several tests with increasing `hashesPerEndpoint` and verify that `stdDev` decreases. 
This would require configurable `hashesPerEndpoint` though.

Reply:
It shouldn't be flaky since the same endpoints are used every time and the hash function doesn't rely on timestamps or anything. But yeah I think this test could be better and test more endpoints and values. I'll see if I can make a reliable test with a low std dev

Reply:
My idea was to use relative measures to demonstrate distribution improves with increasing `hashesPerEndpoint`.
> This would require configurable hashesPerEndpoint though.

That should not necessarily be exposed to the user, we may introduce parameterized constructor for the test and hardcode 100 in algorithm constructor

Reply:
For the series of tests we may also print stdDev to see how it changes in relation to hashesPerEndpoint which could help us selecting its value.

Reply:
For the existing endpoint list (where only the port is changing) I got these results:
```
    algorithm_test.go:444: std deviation with 1 vnodes is 76.562654
    algorithm_test.go:444: std deviation with 2 vnodes is 76.822028
    algorithm_test.go:444: std deviation with 3 vnodes is 77.912749
    algorithm_test.go:444: std deviation with 4 vnodes is 77.170599
    algorithm_test.go:444: std deviation with 5 vnodes is 78.327328
    algorithm_test.go:444: std deviation with 6 vnodes is 78.074813
    algorithm_test.go:444: std deviation with 7 vnodes is 78.707359
    algorithm_test.go:444: std deviation with 8 vnodes is 76.806994
    algorithm_test.go:444: std deviation with 9 vnodes is 66.214453
    algorithm_test.go:444: std deviation with 10 vnodes is 66.045446
    algorithm_test.go:444: std deviation with 11 vnodes is 49.513202
    algorithm_test.go:444: std deviation with 12 vnodes is 48.527946
    algorithm_test.go:444: std deviation with 13 vnodes is 48.401028
    algorithm_test.go:444: std deviation with 14 vnodes is 48.765875
    algorithm_test.go:444: std deviation with 15 vnodes is 48.225465
    algorithm_test.go:444: std deviation with 16 vnodes is 48.569140
    algorithm_test.go:444: std deviation with 17 vnodes is 48.580576
    algorithm_test.go:444: std deviation with 18 vnodes is 48.597207
    algorithm_test.go:444: std deviation with 19 vnodes is 48.675962
    algorithm_test.go:444: std deviation with 20 vnodes is 48.725934
    algorithm_test.go:444: std deviation with 21 vnodes is 36.779492
    algorithm_test.go:444: std deviation with 22 vnodes is 37.690280
    algorithm_test.go:444: std deviation with 23 vnodes is 38.463082
    algorithm_test.go:444: std deviation with 24 vnodes is 39.508029
    algorithm_test.go:444: std deviation with 25 vnodes is 39.501030
    algorithm_test.go:444: std deviation with 26 vnodes is 40.945016
    algorithm_test.go:444: std deviation with 27 vnodes is 41.976303
    algorithm_test.go:444: std deviation with 28 vnodes is 42.885266
    algorithm_test.go:444: std deviation with 29 vnodes is 43.675334
    algorithm_test.go:444: std deviation with 30 vnodes is 44.515224
    algorithm_test.go:444: std deviation with 31 vnodes is 45.580428
    algorithm_test.go:444: std deviation with 32 vnodes is 46.200785
    algorithm_test.go:444: std deviation with 33 vnodes is 46.827413
    algorithm_test.go:444: std deviation with 34 vnodes is 47.936892
    algorithm_test.go:444: std deviation with 35 vnodes is 48.635734
    algorithm_test.go:444: std deviation with 36 vnodes is 49.171013
    algorithm_test.go:444: std deviation with 37 vnodes is 49.801012
    algorithm_test.go:444: std deviation with 38 vnodes is 50.585350
    algorithm_test.go:444: std deviation with 39 vnodes is 52.444257
    algorithm_test.go:444: std deviation with 40 vnodes is 52.887426
    algorithm_test.go:444: std deviation with 41 vnodes is 52.058673
    algorithm_test.go:444: std deviation with 42 vnodes is 52.252125
    algorithm_test.go:444: std deviation with 43 vnodes is 50.879232
    algorithm_test.go:444: std deviation with 44 vnodes is 50.762207
    algorithm_test.go:444: std deviation with 45 vnodes is 50.674746
    algorithm_test.go:444: std deviation with 46 vnodes is 50.696692
    algorithm_test.go:444: std deviation with 47 vnodes is 50.358969
    algorithm_test.go:444: std deviation with 48 vnodes is 50.409388
    algorithm_test.go:444: std deviation with 49 vnodes is 50.350566
    algorithm_test.go:444: std deviation with 50 vnodes is 50.422951
    algorithm_test.go:444: std deviation with 51 vnodes is 50.577697
    algorithm_test.go:444: std deviation with 52 vnodes is 50.860405
    algorithm_test.go:444: std deviation with 53 vnodes is 50.875121
    algorithm_test.go:444: std deviation with 54 vnodes is 51.729290
    algorithm_test.go:444: std deviation with 55 vnodes is 51.718901
    algorithm_test.go:444: std deviation with 56 vnodes is 51.677924
    algorithm_test.go:444: std deviation with 57 vnodes is 51.728355
    algorithm_test.go:444: std deviation with 58 vnodes is 51.756076
    algorithm_test.go:444: std deviation with 59 vnodes is 56.076191
    algorithm_test.go:444: std deviation with 60 vnodes is 57.039225
    algorithm_test.go:444: std deviation with 61 vnodes is 56.991154
    algorithm_test.go:444: std deviation with 62 vnodes is 56.675014
    algorithm_test.go:444: std deviation with 63 vnodes is 56.402591
    algorithm_test.go:444: std deviation with 64 vnodes is 56.488703
    algorithm_test.go:444: std deviation with 65 vnodes is 56.738590
    algorithm_test.go:444: std deviation with 66 vnodes is 56.776072
    algorithm_test.go:444: std deviation with 67 vnodes is 56.530953
    algorithm_test.go:444: std deviation with 68 vnodes is 56.108585
    algorithm_test.go:444: std deviation with 69 vnodes is 55.933288
    algorithm_test.go:444: std deviation with 70 vnodes is 55.827342
    algorithm_test.go:444: std deviation with 71 vnodes is 56.307318
    algorithm_test.go:444: std deviation with 72 vnodes is 56.538410
    algorithm_test.go:444: std deviation with 73 vnodes is 56.547735
    algorithm_test.go:444: std deviation with 74 vnodes is 56.576462
    algorithm_test.go:444: std deviation with 75 vnodes is 56.538923
    algorithm_test.go:444: std deviation with 76 vnodes is 56.525459
    algorithm_test.go:444: std deviation with 77 vnodes is 56.603600
    algorithm_test.go:444: std deviation with 78 vnodes is 56.644078
    algorithm_test.go:444: std deviation with 79 vnodes is 56.587523
    algorithm_test.go:444: std deviation with 80 vnodes is 56.565365
    algorithm_test.go:444: std deviation with 81 vnodes is 56.828796
    algorithm_test.go:444: std deviation with 82 vnodes is 56.929325
    algorithm_test.go:444: std deviation with 83 vnodes is 57.010282
    algorithm_test.go:444: std deviation with 84 vnodes is 57.208769
    algorithm_test.go:444: std deviation with 85 vnodes is 57.078283
    algorithm_test.go:444: std deviation with 86 vnodes is 56.969253
    algorithm_test.go:444: std deviation with 87 vnodes is 56.906652
    algorithm_test.go:444: std deviation with 88 vnodes is 56.616820
    algorithm_test.go:444: std deviation with 89 vnodes is 56.225491
    algorithm_test.go:444: std deviation with 90 vnodes is 56.187699
    algorithm_test.go:444: std deviation with 91 vnodes is 56.016381
    algorithm_test.go:444: std deviation with 92 vnodes is 56.026701
    algorithm_test.go:444: std deviation with 93 vnodes is 56.027927
    algorithm_test.go:444: std deviation with 94 vnodes is 55.835187
    algorithm_test.go:444: std deviation with 95 vnodes is 55.801855
    algorithm_test.go:444: std deviation with 96 vnodes is 55.757177
    algorithm_test.go:444: std deviation with 97 vnodes is 55.767516
    algorithm_test.go:444: std deviation with 98 vnodes is 55.740115
    algorithm_test.go:444: std deviation with 99 vnodes is 55.728670
    algorithm_test.go:444: std deviation with 100 vnodes is 55.723629
```
If I switch endpoints to `endpoints := []string{"http://abc.com:9001", "http://def.com9002", "http://ghi.com:9003", "http://jkl.com:9004", "http://mno.com:9005", "http://pqr.com:9006", "http://stu.com:9007", "http://vwx.com:9008", "http://yza.com:9009", "http://sjsjsjsj.com:9010"}`

Then I get 

```
    algorithm_test.go:444: std deviation with 1 vnodes is 67.957122
    algorithm_test.go:444: std deviation with 2 vnodes is 67.957122
    algorithm_test.go:444: std deviation with 3 vnodes is 70.006628
    algorithm_test.go:444: std deviation with 4 vnodes is 70.006628
    algorithm_test.go:444: std deviation with 5 vnodes is 68.466518
    algorithm_test.go:444: std deviation with 6 vnodes is 68.789123
    algorithm_test.go:444: std deviation with 7 vnodes is 71.162542
    algorithm_test.go:444: std deviation with 8 vnodes is 71.459344
    algorithm_test.go:444: std deviation with 9 vnodes is 66.957709
    algorithm_test.go:444: std deviation with 10 vnodes is 59.010672
    algorithm_test.go:444: std deviation with 11 vnodes is 54.848862
    algorithm_test.go:444: std deviation with 12 vnodes is 54.010125
    algorithm_test.go:444: std deviation with 13 vnodes is 53.058703
    algorithm_test.go:444: std deviation with 14 vnodes is 51.825129
    algorithm_test.go:444: std deviation with 15 vnodes is 54.474798
    algorithm_test.go:444: std deviation with 16 vnodes is 52.550213
    algorithm_test.go:444: std deviation with 17 vnodes is 50.595732
    algorithm_test.go:444: std deviation with 18 vnodes is 48.288835
    algorithm_test.go:444: std deviation with 19 vnodes is 45.848772
    algorithm_test.go:444: std deviation with 20 vnodes is 44.654050
    algorithm_test.go:444: std deviation with 21 vnodes is 54.003684
    algorithm_test.go:444: std deviation with 22 vnodes is 54.399966
    algorithm_test.go:444: std deviation with 23 vnodes is 55.030632
    algorithm_test.go:444: std deviation with 24 vnodes is 55.052335
    algorithm_test.go:444: std deviation with 25 vnodes is 54.762221
    algorithm_test.go:444: std deviation with 26 vnodes is 55.041483
    algorithm_test.go:444: std deviation with 27 vnodes is 55.018527
    algorithm_test.go:444: std deviation with 28 vnodes is 55.828293
    algorithm_test.go:444: std deviation with 29 vnodes is 56.733002
    algorithm_test.go:444: std deviation with 30 vnodes is 57.037392
    algorithm_test.go:444: std deviation with 31 vnodes is 47.687474
    algorithm_test.go:444: std deviation with 32 vnodes is 46.532283
    algorithm_test.go:444: std deviation with 33 vnodes is 47.025971
    algorithm_test.go:444: std deviation with 34 vnodes is 47.939543
    algorithm_test.go:444: std deviation with 35 vnodes is 47.767429
    algorithm_test.go:444: std deviation with 36 vnodes is 47.683636
    algorithm_test.go:444: std deviation with 37 vnodes is 47.398874
    algorithm_test.go:444: std deviation with 38 vnodes is 47.150157
    algorithm_test.go:444: std deviation with 39 vnodes is 46.481656
    algorithm_test.go:444: std deviation with 40 vnodes is 45.285931
    algorithm_test.go:444: std deviation with 41 vnodes is 45.339765
    algorithm_test.go:444: std deviation with 42 vnodes is 45.462170
    algorithm_test.go:444: std deviation with 43 vnodes is 45.961982
    algorithm_test.go:444: std deviation with 44 vnodes is 46.215556
    algorithm_test.go:444: std deviation with 45 vnodes is 46.176723
    algorithm_test.go:444: std deviation with 46 vnodes is 46.149981
    algorithm_test.go:444: std deviation with 47 vnodes is 46.114867
    algorithm_test.go:444: std deviation with 48 vnodes is 45.919234
    algorithm_test.go:444: std deviation with 49 vnodes is 43.192509
    algorithm_test.go:444: std deviation with 50 vnodes is 46.043972
    algorithm_test.go:444: std deviation with 51 vnodes is 46.496399
    algorithm_test.go:444: std deviation with 52 vnodes is 46.466466
    algorithm_test.go:444: std deviation with 53 vnodes is 46.585132
    algorithm_test.go:444: std deviation with 54 vnodes is 46.495839
    algorithm_test.go:444: std deviation with 55 vnodes is 46.429856
    algorithm_test.go:444: std deviation with 56 vnodes is 46.327626
    algorithm_test.go:444: std deviation with 57 vnodes is 46.391375
    algorithm_test.go:444: std deviation with 58 vnodes is 46.612332
    algorithm_test.go:444: std deviation with 59 vnodes is 43.588156
    algorithm_test.go:444: std deviation with 60 vnodes is 43.222541
    algorithm_test.go:444: std deviation with 61 vnodes is 39.663500
    algorithm_test.go:444: std deviation with 62 vnodes is 38.588530
    algorithm_test.go:444: std deviation with 63 vnodes is 37.147164
    algorithm_test.go:444: std deviation with 64 vnodes is 36.288785
    algorithm_test.go:444: std deviation with 65 vnodes is 34.898035
    algorithm_test.go:444: std deviation with 66 vnodes is 34.693796
    algorithm_test.go:444: std deviation with 67 vnodes is 34.905401
    algorithm_test.go:444: std deviation with 68 vnodes is 34.499317
    algorithm_test.go:444: std deviation with 69 vnodes is 34.482336
    algorithm_test.go:444: std deviation with 70 vnodes is 34.609018
    algorithm_test.go:444: std deviation with 71 vnodes is 34.644381
    algorithm_test.go:444: std deviation with 72 vnodes is 35.697718
    algorithm_test.go:444: std deviation with 73 vnodes is 35.689934
    algorithm_test.go:444: std deviation with 74 vnodes is 35.681197
    algorithm_test.go:444: std deviation with 75 vnodes is 35.680997
    algorithm_test.go:444: std deviation with 76 vnodes is 35.680847
    algorithm_test.go:444: std deviation with 77 vnodes is 35.679898
    algorithm_test.go:444: std deviation with 78 vnodes is 35.683499
    algorithm_test.go:444: std deviation with 79 vnodes is 35.719045
    algorithm_test.go:444: std deviation with 80 vnodes is 35.787818
    algorithm_test.go:444: std deviation with 81 vnodes is 35.490638
    algorithm_test.go:444: std deviation with 82 vnodes is 35.640289
    algorithm_test.go:444: std deviation with 83 vnodes is 35.712479
    algorithm_test.go:444: std deviation with 84 vnodes is 35.393885
    algorithm_test.go:444: std deviation with 85 vnodes is 35.167402
    algorithm_test.go:444: std deviation with 86 vnodes is 35.670903
    algorithm_test.go:444: std deviation with 87 vnodes is 36.256012
    algorithm_test.go:444: std deviation with 88 vnodes is 36.267648
    algorithm_test.go:444: std deviation with 89 vnodes is 36.875731
    algorithm_test.go:444: std deviation with 90 vnodes is 34.351043
    algorithm_test.go:444: std deviation with 91 vnodes is 33.906341
    algorithm_test.go:444: std deviation with 92 vnodes is 33.254509
    algorithm_test.go:444: std deviation with 93 vnodes is 33.040620
    algorithm_test.go:444: std deviation with 94 vnodes is 33.062287
    algorithm_test.go:444: std deviation with 95 vnodes is 32.276891
    algorithm_test.go:444: std deviation with 96 vnodes is 31.425914
    algorithm_test.go:444: std deviation with 97 vnodes is 30.569810
    algorithm_test.go:444: std deviation with 98 vnodes is 30.358915
    algorithm_test.go:444: std deviation with 99 vnodes is 30.322190
    algorithm_test.go:444: std deviation with 100 vnodes is 30.272913
```

Interestingly, around 20 is the best value in the first set of endpoints, but in the second set the std deviation gets better as the number of vnodes goes up.

Reply:
Would you like to add the variable vnodes test so we can get this numbers any time?
Having that at hand we may also try out more "realistic" endpoint list where IP address (and not the port) changes. Also we may try to prepend index instead of of appending.

Reply:
Yeah I'll add the test with vnode param. There's lots of things we can vary. I ran a few tests with randomly generated endpoint lists of length 10, and it seemed to get good around 50 vnodes and only get slowly better after that. 

But we would need to vary:
- number of endpoints
- endpoint addresses
- vnode count
- different hashing algorithms
- different ways of creating the vnode key

Reply:
> There's lots of things we can vary.
> it seemed to get good around 50 vnodes

The request keys also has a pattern

Reply:
Actually I've just thought of a better way to measure it. I'll sum up the number of locations on the ring that belong to each endpoint instead. That eliminates the request part of it and just checks how much of the ring is "owned" by each endpoint. Can implement it later today when I finished a few meetings

Reply:
Redid the measurements with the endpoint list `endpoints := []string{"http://10.2.0.1:8080", "http://10.2.0.2:8080", "http://10.2.0.3:8080", "http://10.2.0.4:8080", "http://10.2.0.5:8080", "http://10.2.0.6:8080", "http://10.2.0.7:8080", "http://10.2.0.8:8080", "http://10.2.0.9:8080", "http://10.2.0.10:8080"}`

```
    algorithm_test.go:420: For 1 vnode, standard deviation is 97.180274
    algorithm_test.go:420: For 2 vnode, standard deviation is 96.762329
    algorithm_test.go:420: For 3 vnode, standard deviation is 97.451549
    algorithm_test.go:420: For 4 vnode, standard deviation is 98.978994
    algorithm_test.go:420: For 5 vnode, standard deviation is 96.669586
    algorithm_test.go:420: For 6 vnode, standard deviation is 96.424687
    algorithm_test.go:420: For 7 vnode, standard deviation is 96.856892
    algorithm_test.go:420: For 8 vnode, standard deviation is 97.116022
    algorithm_test.go:420: For 9 vnode, standard deviation is 93.970984
    algorithm_test.go:420: For 10 vnode, standard deviation is 94.516799
    algorithm_test.go:420: For 11 vnode, standard deviation is 52.442926
    algorithm_test.go:420: For 12 vnode, standard deviation is 51.794887
    algorithm_test.go:420: For 13 vnode, standard deviation is 51.552172
    algorithm_test.go:420: For 14 vnode, standard deviation is 53.976798
    algorithm_test.go:420: For 15 vnode, standard deviation is 47.610378
    algorithm_test.go:420: For 16 vnode, standard deviation is 48.255833
    algorithm_test.go:420: For 17 vnode, standard deviation is 46.726952
    algorithm_test.go:420: For 18 vnode, standard deviation is 45.971635
    algorithm_test.go:420: For 19 vnode, standard deviation is 42.696784
    algorithm_test.go:420: For 20 vnode, standard deviation is 38.133112
    algorithm_test.go:420: For 21 vnode, standard deviation is 31.578852
    algorithm_test.go:420: For 22 vnode, standard deviation is 32.794424
    algorithm_test.go:420: For 23 vnode, standard deviation is 33.176572
    algorithm_test.go:420: For 24 vnode, standard deviation is 34.260391
    algorithm_test.go:420: For 25 vnode, standard deviation is 36.781057
    algorithm_test.go:420: For 26 vnode, standard deviation is 37.754357
    algorithm_test.go:420: For 27 vnode, standard deviation is 37.909996
    algorithm_test.go:420: For 28 vnode, standard deviation is 38.240622
    algorithm_test.go:420: For 29 vnode, standard deviation is 43.573172
    algorithm_test.go:420: For 30 vnode, standard deviation is 44.240726
    algorithm_test.go:420: For 31 vnode, standard deviation is 42.899454
    algorithm_test.go:420: For 32 vnode, standard deviation is 42.891375
    algorithm_test.go:420: For 33 vnode, standard deviation is 43.908927
    algorithm_test.go:420: For 34 vnode, standard deviation is 44.318993
    algorithm_test.go:420: For 35 vnode, standard deviation is 45.588943
    algorithm_test.go:420: For 36 vnode, standard deviation is 45.867955
    algorithm_test.go:420: For 37 vnode, standard deviation is 46.294229
    algorithm_test.go:420: For 38 vnode, standard deviation is 46.715923
    algorithm_test.go:420: For 39 vnode, standard deviation is 40.952400
    algorithm_test.go:420: For 40 vnode, standard deviation is 41.512251
    algorithm_test.go:420: For 41 vnode, standard deviation is 40.348957
    algorithm_test.go:420: For 42 vnode, standard deviation is 40.955272
    algorithm_test.go:420: For 43 vnode, standard deviation is 40.822228
    algorithm_test.go:420: For 44 vnode, standard deviation is 41.100913
    algorithm_test.go:420: For 45 vnode, standard deviation is 40.680667
    algorithm_test.go:420: For 46 vnode, standard deviation is 40.725759
    algorithm_test.go:420: For 47 vnode, standard deviation is 40.514179
    algorithm_test.go:420: For 48 vnode, standard deviation is 40.453902
    algorithm_test.go:420: For 49 vnode, standard deviation is 42.023098
    algorithm_test.go:420: For 50 vnode, standard deviation is 41.882094
    algorithm_test.go:420: For 51 vnode, standard deviation is 39.940342
    algorithm_test.go:420: For 52 vnode, standard deviation is 40.064095
    algorithm_test.go:420: For 53 vnode, standard deviation is 41.268490
    algorithm_test.go:420: For 54 vnode, standard deviation is 41.566148
    algorithm_test.go:420: For 55 vnode, standard deviation is 41.949194
    algorithm_test.go:420: For 56 vnode, standard deviation is 42.035254
    algorithm_test.go:420: For 57 vnode, standard deviation is 42.239174
    algorithm_test.go:420: For 58 vnode, standard deviation is 42.451595
    algorithm_test.go:420: For 59 vnode, standard deviation is 42.688479
    algorithm_test.go:420: For 60 vnode, standard deviation is 43.006850
    algorithm_test.go:420: For 61 vnode, standard deviation is 42.998669
    algorithm_test.go:420: For 62 vnode, standard deviation is 42.998252
    algorithm_test.go:420: For 63 vnode, standard deviation is 43.124588
    algorithm_test.go:420: For 64 vnode, standard deviation is 44.426791
    algorithm_test.go:420: For 65 vnode, standard deviation is 44.557691
    algorithm_test.go:420: For 66 vnode, standard deviation is 44.486074
    algorithm_test.go:420: For 67 vnode, standard deviation is 44.504924
    algorithm_test.go:420: For 68 vnode, standard deviation is 44.481851
    algorithm_test.go:420: For 69 vnode, standard deviation is 45.010325
    algorithm_test.go:420: For 70 vnode, standard deviation is 43.907297
    algorithm_test.go:420: For 71 vnode, standard deviation is 44.267156
    algorithm_test.go:420: For 72 vnode, standard deviation is 45.604032
    algorithm_test.go:420: For 73 vnode, standard deviation is 46.780729
    algorithm_test.go:420: For 74 vnode, standard deviation is 47.021814
    algorithm_test.go:420: For 75 vnode, standard deviation is 47.033879
    algorithm_test.go:420: For 76 vnode, standard deviation is 47.045952
    algorithm_test.go:420: For 77 vnode, standard deviation is 47.058032
    algorithm_test.go:420: For 78 vnode, standard deviation is 47.070118
    algorithm_test.go:420: For 79 vnode, standard deviation is 48.647201
    algorithm_test.go:420: For 80 vnode, standard deviation is 32.846622
    algorithm_test.go:420: For 81 vnode, standard deviation is 24.556278
    algorithm_test.go:420: For 82 vnode, standard deviation is 24.179719
    algorithm_test.go:420: For 83 vnode, standard deviation is 22.968245
    algorithm_test.go:420: For 84 vnode, standard deviation is 22.739237
    algorithm_test.go:420: For 85 vnode, standard deviation is 22.115326
    algorithm_test.go:420: For 86 vnode, standard deviation is 22.062013
    algorithm_test.go:420: For 87 vnode, standard deviation is 23.916111
    algorithm_test.go:420: For 88 vnode, standard deviation is 23.738462
    algorithm_test.go:420: For 89 vnode, standard deviation is 23.277906
    algorithm_test.go:420: For 90 vnode, standard deviation is 23.685966
    algorithm_test.go:420: For 91 vnode, standard deviation is 23.363100
    algorithm_test.go:420: For 92 vnode, standard deviation is 23.313394
    algorithm_test.go:420: For 93 vnode, standard deviation is 23.197657
    algorithm_test.go:420: For 94 vnode, standard deviation is 22.824279
    algorithm_test.go:420: For 95 vnode, standard deviation is 22.498074
    algorithm_test.go:420: For 96 vnode, standard deviation is 23.129479
    algorithm_test.go:420: For 97 vnode, standard deviation is 24.038765
    algorithm_test.go:420: For 98 vnode, standard deviation is 23.887029
    algorithm_test.go:420: For 99 vnode, standard deviation is 24.548041
    algorithm_test.go:420: For 100 vnode, standard deviation is 24.559763
```

Reply:
I've switched to the xxhash algorithm, which is already in use in the redis client. It is a lot better:

<details>
  <summary>XXHash</summary>
  
```
    algorithm_test.go:419: For 1 vnode, standard deviation is 103.945546
    algorithm_test.go:419: For 2 vnode, standard deviation is 67.913820
    algorithm_test.go:419: For 3 vnode, standard deviation is 47.120612
    algorithm_test.go:419: For 4 vnode, standard deviation is 42.435186
    algorithm_test.go:419: For 5 vnode, standard deviation is 39.384804
    algorithm_test.go:419: For 6 vnode, standard deviation is 30.566670
    algorithm_test.go:419: For 7 vnode, standard deviation is 37.356923
    algorithm_test.go:419: For 8 vnode, standard deviation is 38.275754
    algorithm_test.go:419: For 9 vnode, standard deviation is 38.548811
    algorithm_test.go:419: For 10 vnode, standard deviation is 30.615408
    algorithm_test.go:419: For 11 vnode, standard deviation is 26.319973
    algorithm_test.go:419: For 12 vnode, standard deviation is 28.929833
    algorithm_test.go:419: For 13 vnode, standard deviation is 25.016320
    algorithm_test.go:419: For 14 vnode, standard deviation is 23.476778
    algorithm_test.go:419: For 15 vnode, standard deviation is 26.581319
    algorithm_test.go:419: For 16 vnode, standard deviation is 26.644824
    algorithm_test.go:419: For 17 vnode, standard deviation is 23.465058
    algorithm_test.go:419: For 18 vnode, standard deviation is 25.301041
    algorithm_test.go:419: For 19 vnode, standard deviation is 20.961722
    algorithm_test.go:419: For 20 vnode, standard deviation is 21.705409
    algorithm_test.go:419: For 21 vnode, standard deviation is 21.097423
    algorithm_test.go:419: For 22 vnode, standard deviation is 22.427356
    algorithm_test.go:419: For 23 vnode, standard deviation is 20.341438
    algorithm_test.go:419: For 24 vnode, standard deviation is 20.415051
    algorithm_test.go:419: For 25 vnode, standard deviation is 20.947251
    algorithm_test.go:419: For 26 vnode, standard deviation is 18.534832
    algorithm_test.go:419: For 27 vnode, standard deviation is 17.527480
    algorithm_test.go:419: For 28 vnode, standard deviation is 13.525875
    algorithm_test.go:419: For 29 vnode, standard deviation is 13.963607
    algorithm_test.go:419: For 30 vnode, standard deviation is 13.809246
    algorithm_test.go:419: For 31 vnode, standard deviation is 14.014038
    algorithm_test.go:419: For 32 vnode, standard deviation is 15.125149
    algorithm_test.go:419: For 33 vnode, standard deviation is 17.663678
    algorithm_test.go:419: For 34 vnode, standard deviation is 15.162547
    algorithm_test.go:419: For 35 vnode, standard deviation is 14.774354
    algorithm_test.go:419: For 36 vnode, standard deviation is 14.014874
    algorithm_test.go:419: For 37 vnode, standard deviation is 13.018295
    algorithm_test.go:419: For 38 vnode, standard deviation is 12.235378
    algorithm_test.go:419: For 39 vnode, standard deviation is 14.397704
    algorithm_test.go:419: For 40 vnode, standard deviation is 16.411218
    algorithm_test.go:419: For 41 vnode, standard deviation is 15.276939
    algorithm_test.go:419: For 42 vnode, standard deviation is 17.539021
    algorithm_test.go:419: For 43 vnode, standard deviation is 16.749588
    algorithm_test.go:419: For 44 vnode, standard deviation is 14.188924
    algorithm_test.go:419: For 45 vnode, standard deviation is 14.323388
    algorithm_test.go:419: For 46 vnode, standard deviation is 13.892631
    algorithm_test.go:419: For 47 vnode, standard deviation is 12.176264
    algorithm_test.go:419: For 48 vnode, standard deviation is 13.197828
    algorithm_test.go:419: For 49 vnode, standard deviation is 12.227785
    algorithm_test.go:419: For 50 vnode, standard deviation is 12.480450
    algorithm_test.go:419: For 51 vnode, standard deviation is 12.433038
    algorithm_test.go:419: For 52 vnode, standard deviation is 12.372138
    algorithm_test.go:419: For 53 vnode, standard deviation is 13.257435
    algorithm_test.go:419: For 54 vnode, standard deviation is 12.760788
    algorithm_test.go:419: For 55 vnode, standard deviation is 13.580368
    algorithm_test.go:419: For 56 vnode, standard deviation is 13.431095
    algorithm_test.go:419: For 57 vnode, standard deviation is 13.317820
    algorithm_test.go:419: For 58 vnode, standard deviation is 14.030707
    algorithm_test.go:419: For 59 vnode, standard deviation is 15.182234
    algorithm_test.go:419: For 60 vnode, standard deviation is 14.873732
    algorithm_test.go:419: For 61 vnode, standard deviation is 15.427417
    algorithm_test.go:419: For 62 vnode, standard deviation is 15.968237
    algorithm_test.go:419: For 63 vnode, standard deviation is 16.520375
    algorithm_test.go:419: For 64 vnode, standard deviation is 16.122582
    algorithm_test.go:419: For 65 vnode, standard deviation is 15.985613
    algorithm_test.go:419: For 66 vnode, standard deviation is 13.392060
    algorithm_test.go:419: For 67 vnode, standard deviation is 13.173599
    algorithm_test.go:419: For 68 vnode, standard deviation is 14.118590
    algorithm_test.go:419: For 69 vnode, standard deviation is 14.258962
    algorithm_test.go:419: For 70 vnode, standard deviation is 13.765616
    algorithm_test.go:419: For 71 vnode, standard deviation is 13.778244
    algorithm_test.go:419: For 72 vnode, standard deviation is 12.048004
    algorithm_test.go:419: For 73 vnode, standard deviation is 11.778376
    algorithm_test.go:419: For 74 vnode, standard deviation is 10.589167
    algorithm_test.go:419: For 75 vnode, standard deviation is 10.862614
    algorithm_test.go:419: For 76 vnode, standard deviation is 12.155013
    algorithm_test.go:419: For 77 vnode, standard deviation is 13.090509
    algorithm_test.go:419: For 78 vnode, standard deviation is 12.896436
    algorithm_test.go:419: For 79 vnode, standard deviation is 12.333919
    algorithm_test.go:419: For 80 vnode, standard deviation is 11.742871
    algorithm_test.go:419: For 81 vnode, standard deviation is 11.616752
    algorithm_test.go:419: For 82 vnode, standard deviation is 12.156079
    algorithm_test.go:419: For 83 vnode, standard deviation is 11.193110
    algorithm_test.go:419: For 84 vnode, standard deviation is 10.959335
    algorithm_test.go:419: For 85 vnode, standard deviation is 11.028586
    algorithm_test.go:419: For 86 vnode, standard deviation is 10.815438
    algorithm_test.go:419: For 87 vnode, standard deviation is 10.442475
    algorithm_test.go:419: For 88 vnode, standard deviation is 10.756830
    algorithm_test.go:419: For 89 vnode, standard deviation is 10.428808
    algorithm_test.go:419: For 90 vnode, standard deviation is 10.706956
    algorithm_test.go:419: For 91 vnode, standard deviation is 10.224502
    algorithm_test.go:419: For 92 vnode, standard deviation is 9.469065
    algorithm_test.go:419: For 93 vnode, standard deviation is 9.053723
    algorithm_test.go:419: For 94 vnode, standard deviation is 8.498528
    algorithm_test.go:419: For 95 vnode, standard deviation is 8.369353
    algorithm_test.go:419: For 96 vnode, standard deviation is 8.145332
    algorithm_test.go:419: For 97 vnode, standard deviation is 7.182943
    algorithm_test.go:419: For 98 vnode, standard deviation is 7.203275
    algorithm_test.go:419: For 99 vnode, standard deviation is 6.701271
    algorithm_test.go:419: For 100 vnode, standard deviation is 5.937805
```

</details>

Reply:
XXHash is 64 bits and FNV-1a is 32-bit so for fair comparison we may also have a look at 64-bit FNV-1a.
But if another hash gives better results then it is worth using.

Reply:
Just tried:

<details>
  <summary>FNV64</summary>
  
```
    algorithm_test.go:418: For 1 vnode, standard deviation is 60.778960
    algorithm_test.go:418: For 2 vnode, standard deviation is 60.778987
    algorithm_test.go:418: For 3 vnode, standard deviation is 60.778961
    algorithm_test.go:418: For 4 vnode, standard deviation is 60.778960
    algorithm_test.go:418: For 5 vnode, standard deviation is 60.778966
    algorithm_test.go:418: For 6 vnode, standard deviation is 60.778964
    algorithm_test.go:418: For 7 vnode, standard deviation is 60.778930
    algorithm_test.go:418: For 8 vnode, standard deviation is 60.778921
    algorithm_test.go:418: For 9 vnode, standard deviation is 60.778954
    algorithm_test.go:418: For 10 vnode, standard deviation is 60.778965
    algorithm_test.go:418: For 11 vnode, standard deviation is 58.852576
    algorithm_test.go:418: For 12 vnode, standard deviation is 58.852546
    algorithm_test.go:418: For 13 vnode, standard deviation is 58.852558
    algorithm_test.go:418: For 14 vnode, standard deviation is 58.852538
    algorithm_test.go:418: For 15 vnode, standard deviation is 58.852542
    algorithm_test.go:418: For 16 vnode, standard deviation is 58.852522
    algorithm_test.go:418: For 17 vnode, standard deviation is 58.852508
    algorithm_test.go:418: For 18 vnode, standard deviation is 58.852488
    algorithm_test.go:418: For 19 vnode, standard deviation is 58.852428
    algorithm_test.go:418: For 20 vnode, standard deviation is 58.852396
    algorithm_test.go:418: For 21 vnode, standard deviation is 58.817270
    algorithm_test.go:418: For 22 vnode, standard deviation is 58.817288
    algorithm_test.go:418: For 23 vnode, standard deviation is 58.817248
    algorithm_test.go:418: For 24 vnode, standard deviation is 58.817248
    algorithm_test.go:418: For 25 vnode, standard deviation is 58.817249
    algorithm_test.go:418: For 26 vnode, standard deviation is 58.817264
    algorithm_test.go:418: For 27 vnode, standard deviation is 58.817222
    algorithm_test.go:418: For 28 vnode, standard deviation is 58.817222
    algorithm_test.go:418: For 29 vnode, standard deviation is 58.817317
    algorithm_test.go:418: For 30 vnode, standard deviation is 58.817338
    algorithm_test.go:418: For 31 vnode, standard deviation is 58.836338
    algorithm_test.go:418: For 32 vnode, standard deviation is 58.836340
    algorithm_test.go:418: For 33 vnode, standard deviation is 58.836379
    algorithm_test.go:418: For 34 vnode, standard deviation is 58.836382
    algorithm_test.go:418: For 35 vnode, standard deviation is 58.836393
    algorithm_test.go:418: For 36 vnode, standard deviation is 58.836396
    algorithm_test.go:418: For 37 vnode, standard deviation is 58.836407
    algorithm_test.go:418: For 38 vnode, standard deviation is 58.836410
    algorithm_test.go:418: For 39 vnode, standard deviation is 58.836491
    algorithm_test.go:418: For 40 vnode, standard deviation is 58.836494
    algorithm_test.go:418: For 41 vnode, standard deviation is 58.767499
    algorithm_test.go:418: For 42 vnode, standard deviation is 58.767507
    algorithm_test.go:418: For 43 vnode, standard deviation is 58.767489
    algorithm_test.go:418: For 44 vnode, standard deviation is 58.767493
    algorithm_test.go:418: For 45 vnode, standard deviation is 58.767504
    algorithm_test.go:418: For 46 vnode, standard deviation is 58.767508
    algorithm_test.go:418: For 47 vnode, standard deviation is 58.767508
    algorithm_test.go:418: For 48 vnode, standard deviation is 58.767508
    algorithm_test.go:418: For 49 vnode, standard deviation is 58.767430
    algorithm_test.go:418: For 50 vnode, standard deviation is 58.767427
    algorithm_test.go:418: For 51 vnode, standard deviation is 58.773321
    algorithm_test.go:418: For 52 vnode, standard deviation is 58.773321
    algorithm_test.go:418: For 53 vnode, standard deviation is 58.773329
    algorithm_test.go:418: For 54 vnode, standard deviation is 58.773329
    algorithm_test.go:418: For 55 vnode, standard deviation is 58.773329
    algorithm_test.go:418: For 56 vnode, standard deviation is 58.773329
    algorithm_test.go:418: For 57 vnode, standard deviation is 58.773329
    algorithm_test.go:418: For 58 vnode, standard deviation is 58.773329
    algorithm_test.go:418: For 59 vnode, standard deviation is 58.773305
    algorithm_test.go:418: For 60 vnode, standard deviation is 58.773305
    algorithm_test.go:418: For 61 vnode, standard deviation is 58.743168
    algorithm_test.go:418: For 62 vnode, standard deviation is 58.743172
    algorithm_test.go:418: For 63 vnode, standard deviation is 58.743130
    algorithm_test.go:418: For 64 vnode, standard deviation is 58.743130
    algorithm_test.go:418: For 65 vnode, standard deviation is 58.743142
    algorithm_test.go:418: For 66 vnode, standard deviation is 58.743146
    algorithm_test.go:418: For 67 vnode, standard deviation is 58.743146
    algorithm_test.go:418: For 68 vnode, standard deviation is 58.743146
    algorithm_test.go:418: For 69 vnode, standard deviation is 58.743043
    algorithm_test.go:418: For 70 vnode, standard deviation is 58.743050
    algorithm_test.go:418: For 71 vnode, standard deviation is 58.749177
    algorithm_test.go:418: For 72 vnode, standard deviation is 58.749180
    algorithm_test.go:418: For 73 vnode, standard deviation is 58.749190
    algorithm_test.go:418: For 74 vnode, standard deviation is 58.749193
    algorithm_test.go:418: For 75 vnode, standard deviation is 58.749193
    algorithm_test.go:418: For 76 vnode, standard deviation is 58.749193
    algorithm_test.go:418: For 77 vnode, standard deviation is 58.749193
    algorithm_test.go:418: For 78 vnode, standard deviation is 58.749193
    algorithm_test.go:418: For 79 vnode, standard deviation is 58.749235
    algorithm_test.go:418: For 80 vnode, standard deviation is 58.749235
    algorithm_test.go:418: For 81 vnode, standard deviation is 58.746464
    algorithm_test.go:418: For 82 vnode, standard deviation is 58.746485
    algorithm_test.go:418: For 83 vnode, standard deviation is 58.746485
    algorithm_test.go:418: For 84 vnode, standard deviation is 58.746506
    algorithm_test.go:418: For 85 vnode, standard deviation is 58.746410
    algorithm_test.go:418: For 86 vnode, standard deviation is 58.746416
    algorithm_test.go:418: For 87 vnode, standard deviation is 58.746402
    algorithm_test.go:418: For 88 vnode, standard deviation is 58.746409
    algorithm_test.go:418: For 89 vnode, standard deviation is 58.746466
    algorithm_test.go:418: For 90 vnode, standard deviation is 58.746480
    algorithm_test.go:418: For 91 vnode, standard deviation is 58.765245
    algorithm_test.go:418: For 92 vnode, standard deviation is 58.765245
    algorithm_test.go:418: For 93 vnode, standard deviation is 58.765245
    algorithm_test.go:418: For 94 vnode, standard deviation is 58.765245
    algorithm_test.go:418: For 95 vnode, standard deviation is 58.765272
    algorithm_test.go:418: For 96 vnode, standard deviation is 58.765272
    algorithm_test.go:418: For 97 vnode, standard deviation is 58.765272
    algorithm_test.go:418: For 98 vnode, standard deviation is 58.765272
    algorithm_test.go:418: For 99 vnode, standard deviation is 58.765385
    algorithm_test.go:418: For 100 vnode, standard deviation is 58.765385
```

</details>

There's a post at https://cyan4973.github.io/xxHash/ which says FNV has poor avalanche properties(i.e. the property of a hash function where changing a small part of it will result in a big change to the hash). This is probably a large part of the reason that my original tests where only the port was changing had 94% of requests going to one backend. It also makes it unsuitable for generating new hashes via appending a number.